pyphare/pyphare/pharein/simulation.py

```diff
@@ -211,41 +212,87 @@ def check_origin(dims, **kwargs):
+
+            if all([isinstance(val, Box) for key,val in boxes.items()]):
+                for box_id, box in boxes.items():
+                    list_per_level[level_key] += [box]
+
+            elif all([isinstance(val, (tuple,list)) for key,val in boxes.items()]):
+                for box_id, box_coords in boxes.items():
+                    box_coords = [list(box_coord) for box_coord in box_coords]
+                    list_per_level[level_key] += [Box(box_coords[0], box_coords[1])]
+
+    return list_per_level
+
+def check_refinement_boxes(dims, **kwargs):
+    """
+      returns tuple ( { "L0" : [Boxes]}, max_nbr_levels)
+    """
+
     refinement_boxes = kwargs.get("refinement_boxes", None)
 
-    if refinement_boxes is None:
-        return refinement_boxes
+    if refinement_boxes is None or len(refinement_boxes) == 0: # SAMRAI errors if empty dict
+        return None, 1 # default max level number to 1 if no boxes
 
     smallest_patch_size = kwargs.get("smallest_patch_size")
+    refinement_ratio = kwargs["refinement_ratio"]
+    nesting_buffer = kwargs["nesting_buffer"]
+
+    refinement_boxes = as_list_per_level(refinement_boxes)
+    domain_box = Box([0] * dims, np.asarray(kwargs["cells"]) - 1)
 
-    for ilvl, boxes in refinement_boxes.items():
+    boxes_per_level = {0: [domain_box]}
 
-        assert ilvl[0] == "L"
+    for level_key, boxes in refinement_boxes.items():
 
-        refined_level_number = int(ilvl[1:])+1
-        if (kwargs["max_nbr_levels"]) <= refined_level_number:
-            err_msg = "Error - refinement boxes to create level {} invalid with 'max_nbr_levels'= {}"
-            raise ValueError(err_msg.format(refined_level_number, kwargs["max_nbr_levels"]))
+        ilvl = int(level_key[1:])
+        refined_level_number = ilvl+1
+        boxes_per_level[refined_level_number] = [boxm.refine(box, refinement_ratio) for box in boxes]
 
-        boxes = list(boxes.values())
         if len(boxes) == 0:
             raise ValueError("Error - missing refinement boxes")
 
-        tmp_boxes = []
-        if isinstance(boxes[0], (tuple,list)):
-            for points in boxes:
-                tmp_boxes.append(Box(points[0], points[1]))
-            boxes = tmp_boxes
+        for box_idx, ref_box in enumerate(boxes):
+            for cmp_box in boxes[box_idx + 1:]:
+                if ref_box * cmp_box != None:
+                    raise ValueError(f"Error: Box({ref_box}) overlaps with Box({cmp_box})")
 
-        elif isinstance(boxes[0], Box):
-            for box in boxes:
-                for l in boxm.refine(box, kwargs["refinement_ratio"]).shape():
-                    if l < smallest_patch_size:
-                        raise  ValueError("Invalid box incompatible with smallest_patch_size")
+        for box in boxes:
+            refined_coarser_boxes = boxes_per_level[ilvl]
 
+            if not any([box in boxm.shrink(refined_coarser, nesting_buffer) for refined_coarser in refined_coarser_boxes]):
```

Comment:
SAMRAI doc says:

```
proper_nesting_buffer A set of max_levels - 1 integer values specifying the number of coarse cells by which the next finer level is nested within the interior of the union of the patches on the next coarser level. The input is given for each level n, where n (= 0, 1,..., N-2) is the level number.
```

therefore I would say we need to shrink the refined_coarser by nesting_buffer*refinement_ratio